rubicon/java/android_events.py

```diff
@@ -0,0 +1,318 @@
+import asyncio
+import asyncio.base_events
+import asyncio.events
+import asyncio.log
+import heapq
+import selectors
+import sys
+import threading
+
+from . import JavaClass, JavaInterface
+
+Handler = JavaClass("android/os/Handler")
+Runnable = JavaInterface("java/lang/Runnable")
+
+
+class AndroidEventLoop(asyncio.SelectorEventLoop):
+    # `AndroidEventLoop` exists to support starting the Python event loop cooperatively with
+    # the built-in Android event loop. Since it's cooperative, it has a `run_forever_cooperatively()`
+    # method which returns immediately. This is is different from the parent class's `run_forever()`,
+    # which blocks.
+    #
+    # In some cases, for simplicity of implementation, this class reaches into the internals of the
+    # parent and grandparent classes.
+    #
+    # A Python event loop needs to handle two things:
+    #
+    # - Waking the event loop when delayed tasks are ready to run.
+    #
+    # - Waking the event loop when I/O is possible on a set of file descriptors.
+    #
+    # `SelectorEventLoop` uses an approach we **cannot** use: it calls the `select()` method
+    # to block waiting for specific file descriptors to be come ready for I/O, or a timeout
+    # corresponding to the soonest delayed task, whichever occurs sooner.
+    #
+    # To handle delayed tasks, `AndroidEventLoop` asks the Android event loop to wake it up when
+    # its soonest delayed task is ready. To accomplish this, it relies on a `SelectorEventLoop`
+    # implementation of `_scheduled` being a collection of tasks sorted by soonest wakeup time.
```

Comment:
Typo: Need to replace `implementation` with `implementation detail`. Will fix ASAP.